mapreduce/src/main/java/uk/gov/gchq/palisade/PalisadeInputFormat.java

```diff
@@ -0,0 +1,391 @@
+            throw new IllegalStateException("No data requests have been registered for this job");
+        }
+        //how many mappers hinted at?
+        int maxMapHint = getMaxMapTasksHint(context);
+        if (maxMapHint < 0) {
+            throw new IllegalStateException("Max map hint illegally set to negative number");
+        }
+
+        //store local and call through method, don't access direct!
+        PalisadeService serv = getPalisadeService();
+
+        ConcurrentLinkedDeque<InputSplit> splits = new ConcurrentLinkedDeque<>();
+        /*Each RegisterDataRequest may result in multiple resources, each of which should be in it's own input split.
+        *These may complete at different rates in any order, so we need to know when all of them have finished to know
+        * that all the resources have been added to the final list. Therefore, we use a Phaser to keep count of the number
+        * of registration requests that have been completed.*/
+        final Phaser counter = new Phaser(1); //register self
+        final int maxCounter = (maxMapHint == 0) ? Integer.MAX_VALUE : maxMapHint;
+        //create a stream for round robining resources
+
+        for (RegisterDataRequest req : reqs) {
+            //tell the phaser we have another party active
+            counter.register();
+            //this iterator determines which mapper gets which resource
+            final PrimitiveIterator.OfInt indexIt = IntStream.iterate(0, x -> (x + 1) % maxCounter).iterator();
+            //send request to the Palisade service
+            serv.registerDataRequest(req)
+                    //when the response comes back create input splits based on max mapper hint
+                    .thenApplyAsync(response -> PalisadeInputFormat.toInputSplits(response, indexIt))
+                            //then add them to the master list (which is thread safe)
+                    .thenAccept(splits::addAll)
+                            //signal this has finished
+                    .whenComplete((r, e) -> counter.arriveAndDeregister());
+        }
+        //wait for everything to finish
+        counter.awaitAdvance(counter.arriveAndDeregister());
+        return new ArrayList<>(splits);
+    }
+
+    /**
+     * Takes a response from the Palisade service and creates a list of input splits. A response may contain many
+     * resources, which we wish to split across multiple input splits. Using the provided iterator, this will
+     * round-robin resources to input splits by grouping them and then making a list of those groups. New {@link
+     * DataRequestResponse}s are created inside the input splits to contain the new distributions.
+     *
+     * @param response the initial response
+     * @param indexIt  the iterator that provides the keys to group the individual resources
+     * @return a list of input splits
+     */
+    public static List<PalisadeInputSplit> toInputSplits(final DataRequestResponse response,
```

Comment:
Could this method could be private? Or are you intending users to use it? It would simplify the public API for the class if we could make it private or move it out into a PalisadeInputFormatUtil class. Similarly with listToMapCollector.

Reply:
Done